docker/docker_exec.go

```diff
@@ -85,92 +87,91 @@ func exec(ctx context.Context, attach, interactive bool, containerName string, a
 	return &insResp, nil
 }
 
-func withRawTerminal(in io.ReadCloser, fn func() error) func() error {
-	return func() error {
-		fd, isTerminal := term.GetFdInfo(in)
-		if isTerminal {
-			var prevState *term.State
-			// set terminal into raw mode to propagate special
-			// characters
-			prevState, err := term.SetRawTerminal(fd)
-			if err != nil {
-				return err
-			}
-
-			defer func() {
-				err = term.RestoreTerminal(fd, prevState)
-			}()
-		}
+type attachFn func(chan<- error)
 
-		return fn()
-	}
+type stdioAttaccher struct {
+	resp *types.HijackedResponse
+	in   io.ReadCloser
+	out  io.Writer
 }
 
-func attachStdin(resp *types.HijackedResponse, in io.ReadCloser, rawTerminal bool) chan error {
-	inputDone := make(chan error)
-
+func (sa *stdioAttaccher) attachStdin(done chan<- error) {
 	go func() {
 		do := func() error {
-			_, err := io.Copy(resp.Conn, in)
+			_, err := io.Copy(sa.resp.Conn, sa.in)
 			if err != nil {
 				return err
 			}
 
-			if err = resp.CloseWrite(); err != nil {
+			if err = sa.resp.CloseWrite(); err != nil {
 				logrus.Debugf("Couldn't send EOF: %s", err)
 			}
 
 			return err
 		}
 
-		if rawTerminal {
-			do = withRawTerminal(in, do)
-		}
-
-		inputDone <- do()
+		done <- do()
 	}()
-
-	return inputDone
 }
 
-func attachStdout(resp *types.HijackedResponse, out io.Writer) chan error {
-	outputDone := make(chan error)
-
+func (sa *stdioAttaccher) attachStdout(done chan<- error) {
 	go func() {
-		_, err := io.Copy(out, resp.Reader)
-		outputDone <- err
-		resp.CloseWrite()
+		_, err := io.Copy(sa.out, sa.resp.Reader)
+		done <- err
+		sa.resp.CloseWrite()
 	}()
-
-	return outputDone
 }
 
-func attachStdio(resp *types.HijackedResponse, in io.ReadCloser, out io.Writer, rawTerminal bool) chan error {
-	done := make(chan error)
+func (sa *stdioAttaccher) attachStdio(done chan<- error) {
 	go func() {
-		do := func() error {
-			inputDone := attachStdin(resp, in, false)
-			outputDone := attachStdout(resp, out)
+		inputDone := make(chan error)
+		outputDone := make(chan error)
+
+		sa.attachStdin(inputDone)
+		sa.attachStdout(outputDone)
+
+		select {
+		case err := <-outputDone:
+			done <- err
+		case err := <-inputDone:
+			if err == nil {
+				// Wait for output to complete streaming.
+				err = <-outputDone
+			}
 
-			select {
-			case err := <-outputDone:
-				return err
-			case err := <-inputDone:
-				if err == nil {
-					// Wait for output to complete streaming.
-					err = <-outputDone
-				}
+			done <- err
+		}
+	}()
+}
 
-				return err
-			}
+func (sa *stdioAttaccher) withRawTerminal(fn attachFn) <-chan error {
+	fd, isTerminal := term.GetFdInfo(sa.in)
+	var err error
+	var prevState *term.State
+	done := make(chan error)
+	wrappedDone := make(chan error)
+	if isTerminal {
+		// set terminal into raw mode to propagate special
+		// characters
+		prevState, err = term.SetRawTerminal(fd)
+		if err != nil {
+			done <- err
```

Comment:
you have deadlock here

Reply:
😮 eagle's eye! thanks for noticing

Reply:
Addressed [here](https://github.com/src-d/engine/pull/439/commits/c39c87b4c41dca3d963944406af51c9f1db04ab0)